src/http/mapping_rule/escaping.rs

```diff
@@ -0,0 +1,262 @@
+            }
+
+            acc
+        })
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    mod coalesce_chars {
+        use super::*;
+
+        #[test]
+        fn remove_duped_forward_slashes() -> Result<(), Error> {
+            let pattern = "/a//b///c////d/////e/";
+            assert_eq!(coalesce_chars(pattern, '/').as_str(), "/a/b/c/d/e/");
+
+            Ok(())
+        }
+    }
+
+    mod split_path_n_qs {
+        use super::*;
+
+        #[test]
+        fn strips_question_mark() -> Result<(), Error> {
+            let pattern = "/abc?param1=1&param2=2";
+            let (path, qs) = split_path_n_qs(pattern);
+
+            assert_eq!(path, "/abc");
+            assert_eq!(qs, Some("param1=1&param2=2"));
+
+            Ok(())
+        }
+    }
+
+    mod query_string_regex {
+        use super::*;
+
+        #[test]
+        fn match_qs_parameters() -> Result<(), Error> {
+            let qs_patterns = "fmt={fmt}&hardcoded=1&lang{num}={lang}";
+            let regexes = query_string_regex(qs_patterns)?;
+            let regexes = regexes.iter().map(Regex::as_str).collect::<Vec<_>>();
+
+            assert!(regexes.contains(&format!("fmt={}", QS_VALUE_REGEX_S).as_str()));
+            assert!(regexes.contains(&format!("hardcoded=1").as_str()));
+            assert!(regexes.contains(&format!("lang{r}={r}", r = QS_VALUE_REGEX_S).as_str()));
+
+            Ok(())
+        }
+    }
+
+    mod path_regex {
+        use super::*;
+
+        #[test]
+        fn match_fail() -> Result<(), Error> {
+            let pattern = "/abc";
+            let regex = path_regex(pattern)?;
+
+            assert!(!regex.is_match("/aaa"));
+
+            Ok(())
+        }
+
+        #[test]
+        fn match_prefix() -> Result<(), Error> {
+            let pattern = "/abc";
+            let regex = path_regex(pattern)?;
+
+            assert!(regex.is_match("/abc"));
+            assert!(regex.is_match("/abcd"));
+
+            Ok(())
+        }
+
+        #[test]
+        fn match_special_chars() -> Result<(), Error> {
+            let pattern = "/foo/{wildcard}/bar";
+            let regex = path_regex(pattern)?;
+
+            assert!(regex.is_match("/foo/a@b/bar"));
+            assert!(regex.is_match("/foo/a:b/bar"));
+            assert!(regex.is_match("/foo/a%b/bar"));
+            assert!(regex.is_match("/foo/a$b/bar"));
+            assert!(regex.is_match("/foo/a()b/bar"));
+
+            Ok(())
+        }
+
+        #[test]
+        fn match_exact() -> Result<(), Error> {
+            let pattern = "/abc$";
+            let regex = path_regex(pattern)?;
+
+            assert!(regex.is_match("/abc"));
+            assert!(!regex.is_match("/abcd"));
+
+            Ok(())
+        }
+
+        #[test]
+        fn match_dollar_sign_at_end() -> Result<(), Error> {
+            let pattern = r"/abc\$";
+            let regex = path_regex(pattern)?;
+
+            assert!(regex.is_match("/abc$"));
+
+            Ok(())
+        }
```

Comment:
This goes against the specification of "using a $ at the end will perform an exact match" (likely because the documentation did not take escaping into account).